experimental/packages/otlp-exporter-base/src/platform/browser/util.ts

```diff
@@ -149,15 +154,72 @@ export function sendWithXhr(
       clearTimeout(retryTimer);
     };
 
-    xhr.onerror = () => {
-      if (reqIsDestroyed) {
-        const err = new OTLPExporterError('Request Timeout');
-        onError(err);
-      }
-      clearTimeout(exporterTimer);
-      clearTimeout(retryTimer);
-    };
+    if (compressionAlgorithm === CompressionAlgorithm.GZIP) {
+
+      const sendRequest = (requestBody: string | Blob) => {
+
+        const send = (body: string | Blob) => {
+          xhr.send(body);
+        };
+
+        const sendCompressed = (body: string | Blob | Uint8Array) => {
+          xhr.setRequestHeader('Content-Encoding', 'gzip'); // Set the Content-Encoding header to 'gzip' for compressed requests
+          xhr.send(body);
+        };
+
+        compressContent(requestBody, compressionAlgorithm)
+          .then(sendCompressed)
+          .catch(() => {
+            send(requestBody); // Send the original body when compression fails
+          });
+      };
+
+      sendRequest(body);
+    } else {
+      xhr.send(body);
+    }
   };
 
   sendWithRetry();
 }
+
+async function compressContent(content: string | Blob, compressionAlgorithm: string): Promise<Uint8Array> {
+
+  const compressionStream = new CompressionStream(compressionAlgorithm as CompressionFormat);
+
+  // Create a new readable stream from the input content
+  const reader = typeof content === 'string' ? new ReadableStream({ start(controller) {
+      controller.enqueue(new TextEncoder().encode(content));
+      controller.close();
+  }}) : content.stream();
+
+  // Pipe the readable stream through the compression stream
+  const compressedStream = reader.pipeThrough(compressionStream);
+
+  // Create a new Uint8Array to hold the compressed data
+  const compressedChunks: Uint8Array[] = [];
+
+  // Read the compressed data from the stream and collect it into chunks
+  const readerStream = compressedStream.getReader();
+  try {
+      while (true) {
+          const { done, value } = await readerStream.read();
+          if (done) break;
+          if (value instanceof Uint8Array) {
+              compressedChunks.push(value);
+          }
+      }
+  } finally {
+      readerStream.releaseLock();
+  }
+
+  // Concatenate the compressed chunks into a single Uint8Array
+  const compressedData = new Uint8Array(compressedChunks.reduce((totalLength, chunk) => totalLength + chunk.length, 0));
```

Comment:
maybe we could get the final size as we're pushing the chunks? Right now we're iterating three times. This would get us down to two.